src/bci_build/registry.py

```diff
@@ -0,0 +1,73 @@
+"""Registry classes for container images."""
+
+import dataclasses
+from abc import ABC
+from abc import abstractmethod
+from dataclasses import dataclass
+
+
+@dataclass(frozen=True, kw_only=True)
+class RegistryABC(ABC):
+    """Abstract Base Class for defining Registry specific content."""
+
+    _: dataclasses.KW_ONLY
+    """The base hostname for this registry instance"""
+    registry: str
+    """The vendor that is put into the ``org.opencontainers.image.vendor`` label"""
+    vendor: str
+
+    @abstractmethod
+    def url(self, container) -> str:
+        pass
+
+    @abstractmethod
+    def registry_prefix(self, *, is_application) -> str:
+        pass
+
+
+class ApplicationCollectionRegistry(RegistryABC):
+    """Registry for the Rancher Application Collection Distribution Platform."""
+
+    def __init__(self):
+        super().__init__(registry="dp.apps.rancher.io", vendor="SUSE LLC")
+
+    def url(self, container) -> str:
+        return f"https://apps.rancher.io/applications/{container.name}"
+
+    def registry_prefix(self, *, is_application) -> str:
+        return "containers"
+
+
+class SUSERegistry(RegistryABC):
+    """Registry for the SUSE Registry."""
+
+    def __init__(self):
+        super().__init__(registry="registry.suse.com", vendor="SUSE LLC")
+
+    def url(self, container) -> str:
+        if container.os_version.is_ltss:
+            return "https://www.suse.com/products/long-term-service-pack-support/"
+        return "https://www.suse.com/products/base-container-images/"
+
+    def registry_prefix(self, *, is_application) -> str:
+        return "suse" if is_application else "bci"
+
+
+class openSUSERegistry(RegistryABC):
+    """Registry for the openSUSE registry."""
+
+    def __init__(self):
+        super().__init__(registry="registry.opensuse.org", vendor="openSUSE Project")
+
+    def url(self, container) -> str:
+        return "https://www.opensuse.org"
+
+    def registry_prefix(self, *, is_application) -> str:
+        return "opensuse" if is_application else "opensuse/bci"
+
+
+def get_registry(container) -> RegistryABC:
+    """Return the appropriate registry for the container."""
+    if container.os_version.is_tumbleweed:
+        return openSUSERegistry()
+    return SUSERegistry()
```

Comment:
Or allow to specify whether this is an appcol image and pass the `os_version` as a parameter instead of the container, that would allow to remove the duplication of `None if os_version.is_tumbleweed else ApplicationCollectionRegistry()`
```suggestion
def get_registry(os_version: OsVersion, is_app_col: bool = False) -> RegistryABC:
    """Return the appropriate registry for the container."""
    if os_version.is_tumbleweed:
        return openSUSERegistry()
    return ApplicationCollectionRegistry() if is_app_col else SUSERegistry()
```

Reply:
this was intentionally the way it is to reduce refactoring effort. using os_version means there are recursive imports, which requires moving code around which you usually object to..

Reply:
splitted into a new toplevel module.